tests/test_config.py

```diff
@@ -174,3 +175,57 @@ def test_requierd_nothing(self):
                 settings={},
                 env={},
             )
+
+
+class TestVerbosity(unittest.TestCase):
+    """Tests for the verbosity configuration from caramel.config"""
+
+    verbosity_data = (
+        (2, {}, logging.CRITICAL, logging.INFO),
+        (-1, {"CARAMEL_LOG_LEVEL": -1}, logging.CRITICAL, logging.ERROR),
+        (2, {"CARAMEL_LOG_LEVEL": 1}, logging.ERROR, logging.INFO),
+        (2, {"CARAMEL_LOG_LEVEL": 3}, logging.WARNING, logging.DEBUG),
+        (2, {"CARAMEL_LOG_LEVEL": 1}, logging.DEBUG, logging.DEBUG),
+        (2, {"CARAMEL_LOG_LEVEL": 3}, logging.WARNING, logging.DEBUG),
+        (0, {"CARAMEL_LOG_LEVEL": 2}, logging.DEBUG, logging.DEBUG),
+        (0, {"CARAMEL_LOG_LEVEL": 3}, logging.WARNING, logging.DEBUG),
+    )  # type: tuple
+    """Data set used for testing config.get_log_level,
+    (argument_level, environment, root_level, expected)"""
+
+    verbosity_argument_data = (
+        ([], 0),
+        (["-v"], 1),
+        (["-vv"], 2),
+        (["-vvv"], 3),
+        (["-vvvv"], 4),
+        (["--verbose", "-v"], 2),
+        (["--verbose", "-vv"], 3),
+        (["--verbose", "-v", "--verbose"], 3),
+    )  # type: tuple
+    """Data set used for testing config.add_verbosity_argument,
+    ([arguments], expected)"""
+
+    def test_verbosity_argument(self):
+        """tests multiple cases for config.add_verbosity_argument"""
+        data = TestVerbosity.verbosity_argument_data
+
+        for arg, expected in data:
+            with self.subTest(arg=arg, expected=expected):
+                parser = argparse.ArgumentParser()
+                config.add_verbosity_argument(parser)
+                args = parser.parse_args(arg)
+                self.assertEqual(expected, args.verbose)
+
+    def test_get_log_level(self):
+        """tests multiple cases for config.get_log_level"""
+        fake_root = logging.getLogger("fake")
+        data = TestVerbosity.verbosity_data
+
+        for arg_lvl, env, root_lvl, expected in data:
```

Comment:
The formulation here suggests that  "data" should be of the form "inputs, expected"

where `inputs` is a dict of the format:

```
{ "arg_lvl": ANY,
  "env": {},
   "root_lvl": ANY,
}
```
Which would make the lines look:


```
for kws, expected in self.VERBOSITY_DATA:
    with self.subTest(**kws, expected=expected)
```
( Note, this is merely commentary, and doesnt need to be adjusted for merge.)

--- 

See also that the function call  here with  "fake_root" reading badly, if it is simply named "logger",  it makes the code read better.